tooling/miniutil/src/build/mod.rs

```diff
@@ -39,6 +39,136 @@ pub use ty::*;
+    pub fn new() -> ProgramBuilder {
+        ProgramBuilder {
+            functions: Default::default(),
+            globals: Default::default(),
+            start: None,
+            next_fn: 0,
+            next_global: 0,
+        }
+    }
+
+    pub fn finish_program(self) -> Program {
+        let start = match self.start {
+            // TODO: decide if default start function 0 is good idea
+            None => FnName(Name::from_internal(0)),
+            Some(start) => start,
+        };
+        Program { functions: self.functions, start: start, globals: self.globals }
+    }
+
+    pub fn function(&mut self) -> FunctionBuilder {
+        let name = FnName(Name::from_internal(self.next_fn));
+        self.next_fn += 1;
+        FunctionBuilder::new(name)
+    }
+
+    pub fn finish_function(&mut self, f: FunctionBuilder) -> FnName {
+        let name = f.get_function_name();
+        let f = f.finish_function();
+        self.functions.try_insert(name, f).unwrap();
+        name
+    }
+
+    pub fn set_start(&mut self, start: FnName) {
+        if self.start.is_some() {
+            panic!("Start function already set");
+        }
+        self.start = Some(start);
+    }
+}
+
+pub struct FunctionBuilder {
+    name: FnName,
+    locals: Map<LocalName, Type>,
+    args: List<LocalName>,
+    blocks: Map<BbName, BasicBlock>,
+
+    // information about current BasicBlock
+    statements: List<Statement>,
+    terminator: Option<Terminator>,
+    block_name: BbName,
+
+    next_block: u32,
+    next_local: u32,
+}
+
+impl FunctionBuilder {
+    fn new(name: FnName) -> FunctionBuilder {
+        FunctionBuilder {
+            name: name,
+            locals: Default::default(),
+            blocks: Default::default(),
+            args: Default::default(),
+            statements: Default::default(),
+            terminator: None,
+            block_name: BbName(Name::from_internal(0)),
+            next_block: 1,
+            // local 0 is always ret
+            next_local: 1,
+        }
+    }
+
+    fn finish_function(mut self) -> Function {
+        // start block and return value are always 0!
+        let start = BbName(Name::from_internal(0));
+        let ret = LocalName(Name::from_internal(0));
+        if self.locals.get(ret).is_none() {
+            self.ret::<()>();
+        }
+        Function {
+            locals: self.locals,
+            args: self.args,
+            ret: ret,
+            calling_convention: CallingConvention::C,
+            blocks: self.blocks,
+            start: start,
+        }
+    }
+
+    pub fn get_function_name(&self) -> FnName {
+        self.name
+    }
+
+    pub fn local<T: TypeConv>(&mut self) -> PlaceExpr {
+        let name = LocalName(Name::from_internal(self.next_local));
+        self.next_local += 1;
+        self.locals.try_insert(name, T::get_type()).unwrap();
+        local_by_name(name)
+    }
+
+    pub fn local_non_const_ty(&mut self, t: Type) -> PlaceExpr {
```

Comment:
The name is confusing. This has nothing to do with `const`, does it?

Maybe `local_with_ty`?